### ISSUE TITLE
[Rotary Embeddings] Interaction of Rotary embeddings for tensor sharding

### ISSUE THREAD

**Issue by codedecde** [facebookresearch/xformers#572]
# ❓ Questions and Help
Hi 
I had a basic question regarding the implementation of Rotary embeddings. 
Currently, the embeddings are split as [hdim // 2], and then combined with the cosine / sine components ([this line](https://github.com/facebookresearch/xformers/blob/bcb707576c6a80eaf850aa80e8643d3497ec2bc4/xformers/components/positional_embedding/rotary.py#L16)). Specifically, the following part

```[python]
def rotate_half(x):
    x1, x2 = x.chunk(2, dim=-1)
    return torch.cat((-x2, x1), dim=-1)


@torch.jit.script
def apply_rotary_pos_emb(x, cos, sin):
    # NOTE: This could probably be moved to Triton

    # Handle a possible sequence length mismatch in between q and k
    cos = cos[:, :, : x.shape[-2], :]
    sin = sin[:, :, : x.shape[-2], :]

    return (x * cos) + (rotate_half(x) * sin)
```

My question is as follows:
Say we train a model with Megatron style sharding, with MP=2. For the query (q) of sequence length T and hidden dim d (assume batch size and heads to be 1 and 1), this would result in
```

q[T, 0:d//2] = [q[T, 0:d//4] * cos - q[T, d//4: d//2] * sin, q[T, d//4:d//2] * cos + q[T, 0:d//4] * sin]  # Device 0
q[T, d//2:] = [q[T, d//2 : 3d//4] * cos - q[T, 3d//4: d] * sin, q[T, 3d//4: d] * cos + q[T, d//2: 3d//4] * sin]  # Device 1
```
Now say we want to change the MP size to 4. The output would be as follows
```
q[T, 0:d//4] = [q[T, 0:d//8] * cos - q[T, d//8: d//4] * sin, q[T, d//8:d//4] * cos + q[T, 0:d//8] * sin]  # Device 0
q[T, d//4:d//2] = [q[T, d//4:3d//8] * cos - q[T, 3d//8: d//2] * sin, q[T, 3d//8:d//2] * cos + q[T, d//4:3d//8] * sin]  # Device 1
...
```

Wouldn't this lead to issues ?

**Comment by bottler** [comment]
I think the RotaryEmbedding class is designed for sequences which begin from the beginning. It will do the wrong thing if you give it just a later part of the sequence, so you can't apply it after splitting the input along its penultimate dimension.

**Comment by codedecde** [comment]
So the splitting is happening along the dimension axis, not the sequence length axis. This is specifically for a Megatron style splitting ([paper](https://arxiv.org/pdf/1909.08053.pdf)) with varying model parallel degree.
<img width="329" alt="image" src="https://user-images.githubusercontent.com/10473264/206760973-1bb11ffa-cdd3-4e91-a7d6-08a3e82b6325.png">

**Comment by codedecde** [comment]
Nevermind, figured it out. It was a misunderstanding on my part. The splitting happens across the multiple heads, and the per head part remains constant with changing MP sizes. Closing this.
